recipes/lora_finetune_single_device.py

```diff
@@ -501,12 +522,123 @@ def _setup_lr_scheduler(
+        """
+        All data related setup happens here. Currently this recipe only supports
+        Map-style Datasets which fit into memory and an option for random shuffling.
+        Samplers, iterable datasets, and streaming datasets are not supported.
+        """
+        if isinstance(cfg_dataset, ListConfig):
+            datasets = [
+                config.instantiate(single_cfg_dataset, self._tokenizer)
+                for single_cfg_dataset in cfg_dataset
+            ]
+            ds = ConcatDataset(datasets=datasets)
+            packed = False
+        else:
+            ds = config.instantiate(cfg_dataset, self._tokenizer)
+            packed = cfg_dataset.get("packed", False)
+
+        sampler = DistributedSampler(
+            ds,
+            num_replicas=1,
+            rank=0,
+            shuffle=shuffle,
+            seed=0,
+        )
+
+        # Instantiate collate_fn
+        if "left_pad_sequence" in collate_fn:
+            raise RuntimeError("left_pad_sequence collator is only for inference.")
+        collate_fn = _get_component_from_path(collate_fn)
+        collate_fn = (
+            partial(
+                collate_fn,
+                padding_idx=self._tokenizer.pad_id,
+                ignore_idx=self._loss_fn.ignore_index,
+            )
+            if not packed
+            else padded_collate_packed
+        )
+
+        from torchdata.nodes import (
+            Batcher,
+            IterableWrapper,
+            Mapper,
+            ParallelMapper,
+            PinMemory,
+            Prefetcher,
+        )
+
+        if True: # Approach 1, drop in replacement for torch.utils.dataloader
+            node = IterableWrapper(sampler)
+            node = Batcher(node, batch_size, drop_last=True)
+            if num_workers == 0:
+                node = Mapper(
+                    node,
+                    map_fn=self._gen_batch_fn(ds=ds, collate_fn=collate_fn),
+                )
+            else:
+                node = ParallelMapper(
+                    node,
+                    map_fn=self._gen_batch_fn(ds=ds, collate_fn=collate_fn),
+                    num_workers=num_workers,
+                    method=parallel_method,
+                    in_order=True,
+                )
+            if pin_memory:
+                node = PinMemory(node)
+            node = Prefetcher(node, 2)
+
+        else: # Approach 2, more granular/pipelined, easy swap for streaming dataset
+            if num_workers == 0:
+                _Mapper = Mapper
+            else:
+                _Mapper = partial(
+                    ParallelMapper,
+                    num_workers=num_workers,
+                    method=parallel_method,
+                    in_order=True,
+                )
+            # Map style set up =======
+            node = IterableWrapper(sampler)
+            node = _Mapper(node, map_fn=ds._data.__getitem__)
+            # Cut here for Streaming/Iterable dataset instead =====
```

Comment:
Have we tried this out on an HF dataset with streaming = True yet? (I assume it won't work out of the box yet?)

Reply:
see update :D almost out-of-the-box

Reply:
why do we need to wrap with `IterableWrapper` if the underlying dataset is IterableDataset? Also in our case, the underlying dataset would be a HF dataset class

Reply:
Wrapper will make it conform to BaseNode's API so we can hold a pointer to iterator, as well as unify the state management. Subclasses need to define `.iterator()` instead of `.__iter__()`.